File: leetcode/938-RangeSumOfBst/RangeSumOfBst.js
class Node {
    constructor(val, left = null, right = null) {
        this.val = val;
        this.left = left;
        this.right = right
    }
}

class BSF {
    constructor() {
        this.root = null;
    }
    
    add(data){
        const node = this.root;
        // if the tree is empty
        if (node == null) {
            this.root = new Node(data);
            return;
    
        //if the tree is not empty    
        } else{
            // create a recursion function that runs while there are more nodes down the trww
            const searchTree = node => {
                // if the data is lower than node
                if (data < node.val) {
                    if (node.left == null) {
                        node.left = new Node(data);
                        return;
                    } else if(node.left != null) {
                        return searchTree(node.left)
                    }

                // if the data is higher than node    
                } else if(data > node.val){
                    if (node.right == null) {
                        node.right = new Node(data);
                        return;
                    } else if(node.right != null) {
                        return searchTree(node.right);
                    }
                } else{
                    return null;
                }
            }
            return searchTree(node);
        }
    }
}

const myFirstTree = new BSF();
myFirstTree.add(8)
myFirstTree.add(4)
myFirstTree.add(10)
myFirstTree.add(25)
myFirstTree.add(2)


// create a function to sum numbers within the range
const sumWithinRange = (tree, low, high) => {
    // if the tree is empty
    if (!tree) {
        return 0;
    }

    let sum = 0;
    // main case to add up values within range
    if (low <= tree.val && tree.val <= high) {
        sum += tree.val        
    }
    // recursive calls to traverse through the tree
    if (tree.left) {
        sum += sumWithinRange(tree.left, low, high);
    }

    if (tree.right) {
        sum += sumWithinRange(tree.right, low, high);
    }

    return sum
}

const sum = sumWithinRange(myFirstTree.root, 3,8); 
console.log(sum)

File: leetcode/938-RangeSumOfBst/RangeSumOfBst.py
# create a node class
class Node:
    def __init__(self, val):
        self.val = val
        self.right = None
        self.left = None

# create a bst class
class Bst:
    def __init__(self):
        self.root = None
        
    # create an add node method within the best class
    def add(self, num):
        # if the bst tree is empty add the new node as the root
        if not self.root:
            self.root = Node(num)
        else:
            # create a function that will be recursive
            def search_tree(node):
                # if the num is less than the tree add it to the right
                if num < node.val:
                    if not node.left:
                        node.left = Node(num)
                        return f"Node {node.left.val} was created"
                    else:
                        search_tree(node.left)

                # if the num is more than the root add it to the left
                elif(num > node.val):
                    if not node.right:
                        node.right = Node(num)
                        return f"Node {node.right.val} was created"
                    else:
                        search_tree(node.right)
                else:
                    return 
                
            # call the recurside function for the first time
            search_tree(self.root)

#created the first bst        
my_first_bst = Bst()

#added nodes to the bst
my_first_bst.add(5)
my_first_bst.add(3)
my_first_bst.add(10)
my_first_bst.add(20)
my_first_bst.add(1)

print(my_first_bst)

# create a sumwithing range function for bst